From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Sun, 12 Jun 2022 21:43:09 +0200
Subject: [PATCH] combat changes


diff --git a/build.gradle.kts b/build.gradle.kts
index fd39ed209b20c927054b8482c400beeeeab460a3..6298692b62df945d74b848c77eec40ea69cdea72 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -94,7 +94,7 @@ sourceSets {
 }
 // Paper end
 // Paper start - brigadier API
-val outgoingVariants = arrayOf("runtimeElements", "apiElements", "sourcesElements", "javadocElements")
+val outgoingVariants = arrayOf("runtimeElements", "apiElements", "sourcesElements") // Papyrus
 configurations {
     val outgoing = outgoingVariants.map { named(it) }
     for (config in outgoing) {
diff --git a/src/main/java/org/bukkit/attribute/Attribute.java b/src/main/java/org/bukkit/attribute/Attribute.java
index 947874c0172b690e7752e49b7bec64e0c0308515..50199c1879152a5c4e272a09201f4d2a91e8d3c0 100644
--- a/src/main/java/org/bukkit/attribute/Attribute.java
+++ b/src/main/java/org/bukkit/attribute/Attribute.java
@@ -6,6 +6,7 @@ import org.bukkit.NamespacedKey;
 import org.bukkit.Translatable;
 import org.jetbrains.annotations.NotNull;
 
+
 /**
  * Types of attributes which may be present on an {@link Attributable}.
  */
@@ -134,7 +135,31 @@ public enum Attribute implements Keyed, Translatable, net.kyori.adventure.transl
     /**
      * Chance of a zombie to spawn reinforcements.
      */
-    ZOMBIE_SPAWN_REINFORCEMENTS("zombie.spawn_reinforcements");
+    ZOMBIE_SPAWN_REINFORCEMENTS("zombie.spawn_reinforcements"),
+
+    // Papyrus start
+    RES_PURE("resistance.pure"),
+    RES_PHYSICAL("resistance.physical"),
+    RES_MAGIC("resistance.magic"),
+    RES_AIR("resistance.air"),
+    RES_EARTH("resistance.earth"),
+    RES_FIRE("resistance.fire"),
+    RES_WATER("resistance.water"),
+
+    ADV_PURE("advantage.pure"),
+    ADV_PHYSICAL("advantage.physical"),
+    ADV_MAGIC("advantage.magic"),
+    ADV_AIR("advantage.air"),
+    ADV_EARTH("advantage.earth"),
+    ADV_FIRE("advantage.fire"),
+    ADV_WATER("advantage.water"),
+
+    STAT_INT("stat.intelligence"),
+    STAT_AGI("stat.agility"),
+    STAT_STR("stat.strength"),
+    STAT_CDR("stat.cooldownreduction"),
+
+    COMBAT_HURTINVUL("combat.hurtinvul");
 
     private final NamespacedKey key;
 
