From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Wed, 26 Apr 2023 21:15:24 +0200
Subject: [PATCH] Profile switching


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/PDamageType.java
similarity index 89%
rename from src/main/java/de/erethon/papyrus/DamageType.java
rename to src/main/java/de/erethon/papyrus/PDamageType.java
index 379e96edce4520a733064044eb357f3041c4ddf0..efa090fd79875852b30e8af27b9bcbc64487b9b1 100644
--- a/src/main/java/de/erethon/papyrus/DamageType.java
+++ b/src/main/java/de/erethon/papyrus/PDamageType.java
@@ -3,7 +3,7 @@ package de.erethon.papyrus;
 
 import org.bukkit.attribute.Attribute;
 
-public enum DamageType {
+public enum PDamageType {
 
     AIR(Attribute.ADV_AIR, Attribute.RES_AIR),
     EARTH(Attribute.ADV_EARTH, Attribute.RES_EARTH),
@@ -16,7 +16,7 @@ public enum DamageType {
     private final Attribute advAttr;
     private final Attribute resAttr;
 
-    DamageType(Attribute advAttr, Attribute resAttr) {
+    PDamageType(Attribute advAttr, Attribute resAttr) {
         this.advAttr = advAttr;
         this.resAttr = resAttr;
     }
diff --git a/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f20fd2c6128cd58ec160e95ca59d276168497a2
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
@@ -0,0 +1,39 @@
+package de.erethon.papyrus;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerSwitchProfileEvent extends PlayerEvent {
+
+    private static final HandlerList handlers = new HandlerList();
+    private int oldProfileID;
+    private int newProfileID;
+
+
+    public PlayerSwitchProfileEvent(@NotNull Player who, int oldProfileID, int newProfileID) {
+        super(who);
+        this.oldProfileID = oldProfileID;
+        this.newProfileID = newProfileID;
+    }
+
+    public int getOldProfileID() {
+        return oldProfileID;
+    }
+
+    public int getNewProfileID() {
+        return newProfileID;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCaster.java b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
index 7260506a0830a69278e6276174857f588c1c6125..3650f4e7206a0de6cf5da42b6d66fd7ceb9cb5ee 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellCaster.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.entity.LivingEntity;
@@ -58,6 +58,10 @@ public interface SpellCaster {
      */
     Set<SpellTrait> getActiveTraits();
 
+    boolean isChanneling();
+
+    void setChanneling(boolean channeling);
+
     default void tick() {
         Iterator<SpellEffect> effectIterator = getEffects().iterator();
         while(effectIterator.hasNext()) {
@@ -82,7 +86,7 @@ public interface SpellCaster {
         }
     }
 
-    default double onDamage(LivingEntity damager, double damage, DamageType type) {
+    default double onDamage(LivingEntity damager, double damage, PDamageType type) {
         for (SpellbookSpell spell : getActiveSpells()) {
             damage = spell.onDamage(damager, damage, type);
         }
@@ -95,7 +99,7 @@ public interface SpellCaster {
         return Math.max(damage, 0);
     }
 
-    default double onAttack(LivingEntity target, double damage, DamageType type) {
+    default double onAttack(LivingEntity target, double damage, PDamageType type) {
         for (SpellbookSpell spell : getActiveSpells()) {
             damage = spell.onAttack(target, damage, type);
         }
@@ -318,5 +322,26 @@ public interface SpellCaster {
         }
     }
 
+    /**
+     * This will only interrupt channeling spells for now.
+     */
+    default void interrupt() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
+    default void interruptFromMovement() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling && spell.isMovementInterrupted) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
 
 }
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3ab73e1736b09e3b1ed053463848993b81259bc0
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
@@ -0,0 +1,41 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelFinishEvent extends Event {
+
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+    private final boolean interrupted;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelFinishEvent(SpellCaster caster, SpellbookSpell spell, boolean interrupted) {
+        this.caster = caster;
+        this.spell = spell;
+        this.interrupted = interrupted;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    public boolean wasInterrupted() {
+        return interrupted;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..ff0571c54408568ef66cc149e10a9daa2528b1d4
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
@@ -0,0 +1,34 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelStartEvent extends Event {
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelStartEvent(SpellCaster caster, SpellbookSpell spell) {
+        this.caster = caster;
+        this.spell = spell;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellData.java b/src/main/java/de/erethon/spellbook/api/SpellData.java
index 1f8dd5a590298a7b0bbb2b47354163eb671cec37..23ca423cc45eb9586b0a25a0a976f4cc8b2b85c2 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellData.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellData.java
@@ -10,8 +10,7 @@ import org.jetbrains.annotations.NotNull;
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
 
 public class SpellData extends YamlConfiguration {
 
@@ -20,8 +19,8 @@ public class SpellData extends YamlConfiguration {
     SpellQueue queue;
     private int cooldown;
     private final String id;
-    private Component name;
-    private final List<String> description = new ArrayList<>();
+    private final Map<Locale, Component> name = new HashMap<>();
+    private final Map<Locale, List<String>> description = new HashMap<>();
 
     private Class<? extends SpellbookSpell> spellClass;
     private final ClassLoader classLoader;
@@ -72,24 +71,24 @@ public class SpellData extends YamlConfiguration {
     }
 
     @NotNull
-    public Component getDisplayName() {
-        return name;
+    public Component getDisplayName(Locale locale) {
+        return name.get(locale);
     }
 
-    public List<Component> getDescription() {
+    public List<Component> getDescription(Locale locale) {
         List<Component> components = new ArrayList<>();
-        for (String entry : description) {
+        for (String entry : description.get(locale)) {
             components.add(miniMessage.deserialize(entry));
         }
         return components;
     }
 
-    public List<Component> getDescription(String[] placeholders, String[] replacements) {
+    public List<Component> getDescription(Locale locale, String[] placeholders, String[] replacements) {
         if (replacements.length < placeholders.length) {
             throw new RuntimeException("Invalid amount of replacements for provided placeholders");
         }
         List<Component> components = new ArrayList<>();
-        for (String entry : description) {
+        for (String entry : description.get(locale)) {
             for (int i = 0; i < placeholders.length; i++) {
                 components.add(miniMessage.deserialize(entry.replaceAll(placeholders[i], replacements[i])));
             }
@@ -121,7 +120,12 @@ public class SpellData extends YamlConfiguration {
             throw new RuntimeException(e);
         }
         cooldown = getInt("cooldown", 0);
-        name = miniMessage.deserialize(getString("name", "<color:#ff0000>[Invalid name]"));
-        description.addAll(getStringList("description"));
+        for (String locale : getConfigurationSection("name").getKeys(false)) {
+            name.put(Locale.of(locale), miniMessage.deserialize(getString("name." + locale, "<color:#ff0000>[Invalid name]")));
+        }
+        for (String locale : getConfigurationSection("description").getKeys(false)) {
+            List<String> entries = getStringList("description." + locale);
+            description.put(Locale.of(locale), entries);
+        }
     }
 }
diff --git a/src/main/java/de/erethon/spellbook/api/SpellEffect.java b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
index 00e3dc889b0c4978fa20a76f233ede7bea34f878..2b28cf882250df1836b5b5b9c8a0a9f9864e1430 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellEffect.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.entity.LivingEntity;
 
 public class SpellEffect {
@@ -47,11 +47,11 @@ public class SpellEffect {
     public void onRemove() { // Mostly useful for stat buffs/debuffs
     }
 
-    public double onDamage(LivingEntity damager, double damage, DamageType type) { // For effects that block 1 attack and get removed after that
+    public double onDamage(LivingEntity damager, double damage, PDamageType type) { // For effects that block 1 attack and get removed after that
         return damage;
     }
 
-    public double onAttack(LivingEntity target, double damage, DamageType type) { // For effects that add something to the attack
+    public double onAttack(LivingEntity target, double damage, PDamageType type) { // For effects that add something to the attack
         return damage;
     }
 
diff --git a/src/main/java/de/erethon/spellbook/api/SpellTrait.java b/src/main/java/de/erethon/spellbook/api/SpellTrait.java
index 96ec2b714f8e9f8c69f26811ee2878d20dfee5b6..828b24e26444f98b423d79a4f2775ea02348f609 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellTrait.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellTrait.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.entity.LivingEntity;
 
 import java.util.Set;
@@ -81,7 +81,7 @@ public class SpellTrait {
      * This method is called when the caster takes damage.
      * @param attacker the attacker
      */
-    public double onDamage(LivingEntity attacker, double damage, DamageType type) {
+    public double onDamage(LivingEntity attacker, double damage, PDamageType type) {
         return damage;
     }
 
@@ -89,7 +89,7 @@ public class SpellTrait {
      * This method is called when the caster deals damage.
      * @param target the target the caster is attacking
      */
-    public double onAttack(LivingEntity target, double damage, DamageType type) {
+    public double onAttack(LivingEntity target, double damage, PDamageType type) {
         return damage;
     }
 
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
index c4a07139e108bbd54e33d499bb883b6ec6eb8baa..e4ccda7c3ffdc025612c41a04c7c72af150d2173 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
@@ -1,8 +1,11 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Bukkit;
 import org.bukkit.entity.LivingEntity;
 
+import java.util.Locale;
 import java.util.Set;
 import java.util.UUID;
 
@@ -16,12 +19,16 @@ public abstract class SpellbookSpell {
 
     protected int keepAliveTicks = 0;
     protected int currentTicks = 0;
+    public int channelDuration = 0;
+    public boolean isMovementInterrupted = false;
+    public int currentChannelTicks = 0;
 
     protected int tickInterval = 1;
     protected int currentTickInterval = 0;
 
     protected boolean failed = false;
     protected boolean interrupted = false;
+    public boolean isChanneling = false;
 
     public SpellbookSpell(LivingEntity caster, SpellData spellData) {
         this.data = spellData;
@@ -67,6 +74,19 @@ public abstract class SpellbookSpell {
         cleanup();
     }
 
+    /**
+     * This is method is called after channeling was finished.
+     */
+    protected void onChannelFinish() {
+    }
+
+    /**
+     * This method is called when channeling is interrupted for whatever reason. cleanup() will be called afterwards
+     */
+    protected void onChannelInterrupt() {
+
+    }
+
     /**
      * This method should be used for cleaning up after a Spell cast. For example spawned entities should be
      * removed here.
@@ -78,7 +98,7 @@ public abstract class SpellbookSpell {
      * This method is called when the caster takes damage.
      * @param attacker the attacker
      */
-    public double onDamage(LivingEntity attacker, double damage, DamageType type) {
+    public double onDamage(LivingEntity attacker, double damage, PDamageType type) {
         return damage;
     }
 
@@ -86,7 +106,7 @@ public abstract class SpellbookSpell {
      * This method is called when the caster deals damage.
      * @param target the target the caster is attacking
      */
-    public double onAttack(LivingEntity target, double damage, DamageType type) {
+    public double onAttack(LivingEntity target, double damage, PDamageType type) {
         return damage;
     }
 
@@ -197,12 +217,27 @@ public abstract class SpellbookSpell {
 
     /**
      * Interrupts the spell. It will stop ticking, but cleanup() will still
-     * be called.
+     * be called. onChannelInterrupt() will be called first if the Spell is currently channeling.
      */
     public void interrupt() {
+        if (isChanneling) {
+            SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, true);
+            Bukkit.getPluginManager().callEvent(finishEvent);
+            onChannelInterrupt();
+            channelDuration = -1; // Don't continue channeling
+        }
         interrupted = true;
     }
 
+    /**
+     * This is useful for placeholders that depend on specific conditions and attributes
+     * @param id the ID
+     * @return the value
+     */
+    public String getValueWithID(int id) {
+        return "<none>";
+    }
+
     /**
      * Immediately stops spell execution.
      * <b>Careful! No other Spell methods with be called afterwards.</b>
@@ -235,6 +270,22 @@ public abstract class SpellbookSpell {
             cleanup();
             return;
         }
+        if (channelDuration > 0) {
+            currentChannelTicks++;
+            if (!isChanneling && currentChannelTicks < channelDuration) {
+                isChanneling = true;
+                caster.setChanneling(true);
+                SpellChannelStartEvent startEvent = new SpellChannelStartEvent(caster, this);
+                Bukkit.getPluginManager().callEvent(startEvent);
+            }
+            if (currentChannelTicks == channelDuration) {
+                SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, false);
+                Bukkit.getPluginManager().callEvent(finishEvent);
+                onChannelFinish();
+                isChanneling = false;
+                caster.setChanneling(false);
+            }
+        }
         currentTicks++;
         if (currentTickInterval >= tickInterval) {
             currentTickInterval = 0;
diff --git a/src/main/java/de/erethon/spellbook/api/TraitData.java b/src/main/java/de/erethon/spellbook/api/TraitData.java
index 5c78b4f736ad271dcc512d1c85f83728d956dc00..c7bd36aeb18a33a444bbd3a67caa45f4fead356c 100644
--- a/src/main/java/de/erethon/spellbook/api/TraitData.java
+++ b/src/main/java/de/erethon/spellbook/api/TraitData.java
@@ -10,10 +10,7 @@ import org.jetbrains.annotations.NotNull;
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 
 public class TraitData extends YamlConfiguration {
 
@@ -26,8 +23,8 @@ public class TraitData extends YamlConfiguration {
     private Class<? extends SpellTrait> traitClass;
     private final ClassLoader classLoader;
 
-    private Component name;
-    private final List<String> description = new ArrayList<>();
+    private Map<Locale, Component> name = new HashMap<>();
+    private final Map<Locale, List<String>> description = new HashMap<>();
     private final Set<SpellData> affectedSpells = new HashSet<>();
     private final Set<TraitData> affectedTraits = new HashSet<>();
 
@@ -50,17 +47,17 @@ public class TraitData extends YamlConfiguration {
     }
 
     @NotNull
-    public Component getDisplayName() {
-        return name;
+    public Component getDisplayName(Locale locale) {
+        return name.get(locale);
     }
 
     /**
      * @return a List of Components
      * <b>Note: For most cases, the method with placeholders should be used.</b>
      */
-    public List<Component> getDescription() {
+    public List<Component> getDescription(Locale locale) {
         List<Component> components = new ArrayList<>();
-        for (String entry : description) {
+        for (String entry : description.get(locale)) {
             components.add(miniMessage.deserialize(entry));
         }
         return components;
@@ -71,12 +68,12 @@ public class TraitData extends YamlConfiguration {
      * @param replacements array of strings to replace those placeholders
      * @return a List of Components with the placeholders replaced.
      */
-    public List<Component> getDescription(String[] placeholders, String[] replacements) {
+    public List<Component> getDescription(Locale locale, String[] placeholders, String[] replacements) {
         if (replacements.length < placeholders.length) {
             throw new RuntimeException("Invalid amount of replacements for provided placeholders");
         }
         List<Component> components = new ArrayList<>();
-        for (String entry : description) {
+        for (String entry : description.get(locale)) {
             for (int i = 0; i < placeholders.length; i++) {
                 components.add(miniMessage.deserialize(entry.replaceAll(placeholders[i], replacements[i])));
             }
@@ -126,8 +123,13 @@ public class TraitData extends YamlConfiguration {
             spellbookAPI.getServer().getLogger().warning("Could not find class for trait " + id + ": " + className);
             throw new RuntimeException(e1);
         }
-        name = miniMessage.deserialize(getString("name", "<color:#ff0000>[Invalid name]"));
-        description.addAll(getStringList("description"));
+        for (String locale : getConfigurationSection("name").getKeys(false)) {
+            name.put(Locale.of(locale), miniMessage.deserialize(getString("name." + locale, "<color:#ff0000>[Invalid name]")));
+        }
+        for (String locale : getConfigurationSection("description").getKeys(false)) {
+            List<String> entries = getStringList("description." + locale);
+            description.put(Locale.of(locale), entries);
+        }
         List<String> affectedIDs = getStringList("affectedSpells");
         for (String id : affectedIDs) {
             SpellData data = spellbookAPI.getLibrary().getSpellByID(id);
diff --git a/src/main/java/org/bukkit/attribute/Attribute.java b/src/main/java/org/bukkit/attribute/Attribute.java
index 8bc1b04386ab2642b74d57d1ee44485c0bea9f92..ace00c147a89bb60ea99962f0fa4a6f60a5b9086 100644
--- a/src/main/java/org/bukkit/attribute/Attribute.java
+++ b/src/main/java/org/bukkit/attribute/Attribute.java
@@ -140,7 +140,7 @@ public enum Attribute implements Keyed, Translatable, net.kyori.adventure.transl
     // Papyrus start
     RES_PURE("resistance.pure"),
     RES_PHYSICAL("resistance.physical"),
-    RES_MAGIC("resistance.magic"),
+    RES_MAGIC("resistance.magical"),
     RES_AIR("resistance.air"),
     RES_EARTH("resistance.earth"),
     RES_FIRE("resistance.fire"),
@@ -148,8 +148,8 @@ public enum Attribute implements Keyed, Translatable, net.kyori.adventure.transl
 
     ADV_PURE("advantage.pure"),
     ADV_PHYSICAL("advantage.physical"),
-    ADV_MAGIC("advantage.magic"),
-    PEN_MAGIC("penetration.magic"),
+    ADV_MAGIC("advantage.magical"),
+    PEN_MAGIC("penetration.magical"),
     PEN_PHYSICAL("penetration.physical"),
     ADV_AIR("advantage.air"),
     ADV_EARTH("advantage.earth"),
diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index a81d5720f19f15dcffdf346f5f8d4e23cfb7a9ee..894c65c76c639205fb8e9e38258cde67b94c64dd 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -1,6 +1,6 @@
 package org.bukkit.entity;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.damage.DamageSource;
 import org.jetbrains.annotations.ApiStatus;
@@ -17,7 +17,7 @@ public interface Damageable extends Entity {
      * @param amount Amount of damage to deal
      */
     void damage(double amount);
-    void damage(double amount, DamageType type);
+    void damage(double amount, PDamageType type);
 
 
     /**
@@ -29,7 +29,7 @@ public interface Damageable extends Entity {
      */
     void damage(double amount, @Nullable Entity source);
 
-    void damage(double amount, Entity source, DamageType type);
+    void damage(double amount, Entity source, PDamageType type);
 
     /**
      * Deals the given amount of damage to this entity from a specified
diff --git a/src/main/java/org/bukkit/entity/Projectile.java b/src/main/java/org/bukkit/entity/Projectile.java
index 9fe73c563e015c37d2056218599ba21f283ae980..8816c19764150ed1dd406c4f98394b4383672d24 100644
--- a/src/main/java/org/bukkit/entity/Projectile.java
+++ b/src/main/java/org/bukkit/entity/Projectile.java
@@ -1,6 +1,6 @@
 package org.bukkit.entity;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.projectiles.ProjectileSource;
 import org.jetbrains.annotations.Nullable;
 
@@ -128,5 +128,5 @@ public interface Projectile extends Entity {
     @Nullable
     java.util.UUID getOwnerUniqueId();
     // Paper end
-    void setDamageType(DamageType type);
+    void setDamageType(PDamageType type);
 }
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
index 148c4aad384ae8e3b8b22d264a84bddfbcafdf1e..90a5f69d513833f116715f2f546593f92a2dcc81 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
@@ -2,6 +2,8 @@ package org.bukkit.event.entity;
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockState;
 import org.bukkit.damage.DamageSource;
@@ -38,7 +40,7 @@ public class EntityDamageByBlockEvent extends EntityDamageEvent {
     }
 
     public EntityDamageByBlockEvent(@Nullable final Block damager, @Nullable final BlockState damagerState, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, PDamageType.PURE);
         this.damager = damager;
         this.damagerState = damagerState;
     }
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
index 8ea4be529400b34df3d31b0f17c2d145345523d9..e9a87df62a86ee6ba3f2cabf702df25121b84ae1 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -2,8 +2,11 @@ package org.bukkit.event.entity;
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.damage.DamageSource;
 import org.bukkit.damage.DamageType;
+
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
@@ -27,24 +30,30 @@ public class EntityDamageByEntityEvent extends EntityDamageEvent {
         this.critical = false; // Paper - add critical damage API
     }
 
-    @Deprecated(forRemoval = true)
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage, DamageType type) {
+        super(damagee, cause, damage);
+        this.damager = damager;
+        this.critical = false; // Paper - add critical damage API
+    }
+
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
         this(damager, damagee, cause, DamageSource.builder(DamageType.GENERIC).withCausingEntity(damager).withDirectEntity(damager).build(), modifiers, modifierFunctions);
     }
 
     @Deprecated
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, PDamageType.PURE);
         this.damager = damager;
         // Paper start
         this.critical = false;
     }
 
     @ApiStatus.Internal
-    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, boolean critical) {
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, boolean critical, PDamageType pType) {
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, pType);
         this.damager = damager;
         this.critical = critical;
+        this.pDamageType = pType;
     }
 
     /**
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index ef5b2a0f18c1c126db0b0c4a4d2a57483680665a..6f7875aa72c3f9c872fe38d364ae5ac8ac36ee9f 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -7,6 +7,8 @@ import com.google.common.collect.ImmutableMap;
 import java.util.EnumMap;
 import java.util.Map;
 import java.util.Objects;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.Material;
 import org.bukkit.WorldBorder;
 import org.bukkit.damage.DamageSource;
@@ -31,21 +33,22 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
     private final DamageCause cause;
     private final DamageSource damageSource;
 
-    @Deprecated(forRemoval = true)
+    protected PDamageType pDamageType; // Papyrus
+
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
         this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), damage);
     }
 
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, final double damage) {
-        this(damagee, cause, damageSource, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)));
+        this(damagee, cause, damageSource, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)), PDamageType.PURE);
     }
 
-    @Deprecated(forRemoval = true)
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
-        this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), modifiers, modifierFunctions);
+        this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), modifiers, modifierFunctions, PDamageType.PURE);
     }
 
-    public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
+
+    public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, DamageSource source, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, PDamageType pType) {
         super(damagee);
         Preconditions.checkArgument(modifiers.containsKey(DamageModifier.BASE), "BASE DamageModifier missing");
         Preconditions.checkArgument(!modifiers.containsKey(null), "Cannot have null DamageModifier");
@@ -56,7 +59,8 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         this.cause = cause;
         this.modifiers = modifiers;
         this.modifierFunctions = modifierFunctions;
-        this.damageSource = damageSource;
+        this.damageSource = source;
+        this.pDamageType = pType; // Papyrus
     }
 
     @Override
@@ -232,6 +236,14 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         return handlers;
     }
 
+    public PDamageType getDamageType() {
+        return pDamageType;
+    }
+
+    public void setDamageType(PDamageType type) {
+        pDamageType = type;
+    }
+
     /**
      * An enum to specify the types of modifier
      *
