From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Sat, 11 Jun 2022 18:07:35 +0200
Subject: [PATCH] combat changes


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/DamageType.java
@@ -0,0 +1,32 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+public enum DamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+
+    DamageType(Attribute advAttr, Attribute resAttr) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/PapyrusUtils.java b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..afe4ed93f622cc46c0bb1faa12976e44297f4ea9
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
@@ -0,0 +1,14 @@
+package de.erethon.papyrus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.network.chat.Component;
+
+public class PapyrusUtils {
+
+    private static final char healthBar = '|';
+
+    public static Component getHealthMessage(double health, double max) { // maybe with color based on damage type?
+        return PaperAdventure.asVanilla(MiniMessage.miniMessage().deserialize("<red>" + Math.round(health) + "<dark_gray>/" + "<dark_red>" + Math.round(max)));
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 3b6945e58b157c89d99fff020003a5b4e67b9e0b..2199ca8112057ecfe32c6301988093b83c13b13b 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -6,6 +6,7 @@ import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import de.erethon.papyrus.DamageType;
 import io.papermc.paper.adventure.PaperAdventure;
 import net.minecraft.*;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -2342,7 +2343,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target);
+            super.attack(target, DamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7796e191747be545e744564a2b0b65790f69114d..47fa5fb1884ee33f03f875951442a27d0df12130 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -314,6 +314,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
         Objects.requireNonNull(server);
         this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(uuid, server::enforceSecureProfile);
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
+        LOGGER.info("Initialized packet listener for " + player.getScoreboardName());
     }
 
     // CraftBukkit start - add fields and methods
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 83f3ffdd8fa901b3de580d2359cdb5ead0d762cb..5212cfad27b6f7dd715dbf69008e6366e031b3da 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -10,6 +10,7 @@ import it.unimi.dsi.fastutil.doubles.DoubleList;
 import it.unimi.dsi.fastutil.doubles.DoubleListIterator;
 import it.unimi.dsi.fastutil.floats.FloatArraySet;
 import it.unimi.dsi.fastutil.floats.FloatArrays;
+import de.erethon.papyrus.DamageType;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.ArrayList;
@@ -2143,6 +2144,16 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
+    // Papyrus - combat rework
+    public boolean hurt(DamageSource source, float amount, DamageType type) {
+        if (this.isInvulnerableTo(source)) {
+            return false;
+        } else {
+            this.markHurt();
+            return false;
+        }
+    }
+
     public final Vec3 getViewVector(float tickDelta) {
         return this.calculateViewVector(this.getViewXRot(tickDelta), this.getViewYRot(tickDelta));
     }
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index da56b9704ce486f4c5956cc044a73b10fd2db140..4930eef744cdb9f1855502a3186235cf483358b9 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -39,6 +39,30 @@ public class Attributes {
     public static final Holder<Attribute> SUBMERGED_MINING_SPEED = Attributes.register("player.submerged_mining_speed", (new RangedAttribute("attribute.name.player.submerged_mining_speed", 0.2D, 0.0D, 20.0D)).setSyncable(true));
     public static final Holder<Attribute> SWEEPING_DAMAGE_RATIO = Attributes.register("player.sweeping_damage_ratio", (new RangedAttribute("attribute.name.player.sweeping_damage_ratio", 0.0D, 0.0D, 1.0D)).setSyncable(true));
     public static final Holder<Attribute> WATER_MOVEMENT_EFFICIENCY = Attributes.register("generic.water_movement_efficiency", (new RangedAttribute("attribute.name.generic.water_movement_efficiency", 0.0D, 0.0D, 1.0D)).setSyncable(true));
+    // Papyrus start - combat changes
+    // Element resistance
+    public static final Holder<Attribute> RES_PURE = Attributes.register("resistance.pure", new RangedAttribute("attribute.name.resistance.pure", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_PHYSICAL = Attributes.register("resistance.physical", new RangedAttribute("attribute.name.resistance.physical", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_MAGIC = Attributes.register("resistance.magic", new RangedAttribute("attribute.name.resistance.magic", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_AIR = Attributes.register("resistance.air", new RangedAttribute("attribute.name.resistance.air", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_EARTH = Attributes.register("resistance.earth", new RangedAttribute("attribute.name.resistance.earth", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_FIRE = Attributes.register("resistance.fire", new RangedAttribute("attribute.name.resistance.fire", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> RES_WATER = Attributes.register("resistance.water", new RangedAttribute("attribute.name.resistance.water", 0, 0, 4096)).setSyncable(false);
+    // Element advantageHolder<Attribute>
+    public static final Holder<Attribute> ADV_PURE = Attributes.register("advantage.pure", new RangedAttribute("attribute.name.advantage.pure", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_PHYSICAL = Attributes.register("advantage.physical", new RangedAttribute("attribute.name.advantage.physical", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_MAGIC = Attributes.register("advantage.magic", new RangedAttribute("attribute.name.advantage.magic", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_AIR = Attributes.register("advantage.air", new RangedAttribute("attribute.name.advantage.air", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_EARTH = Attributes.register("advantage.earth", new RangedAttribute("attribute.name.advantage.earth", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_FIRE = Attributes.register("advantage.fire", new RangedAttribute("attribute.name.advantage.fire", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> ADV_WATER = Attributes.register("advantage.water", new RangedAttribute("attribute.name.advantage.water", 0, 0, 4096)).setSyncable(false);
+    // StatsHolder<Attribute>
+    public static final Holder<Attribute> STAT_CRITCHANCE = Attributes.register("stat.crit.chance", new RangedAttribute("attribute.name.stat.crit.chance", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> STAT_ATTACKSPEED= Attributes.register("stat.attackspeed", new RangedAttribute("attribute.name.stat.attackspeed", 0, 0, 4096)).setSyncable(false);
+    public static final Holder<Attribute> STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, 0, 4096)).setSyncable(false);
+    // CombatHolder<Attribute>
+    public static final Holder<Attribute> COMBAT_HURTINVUL= Attributes.register("combat.hurtinvul", new RangedAttribute("attribute.name.combat.hurtinvul", 5, 0, 4096)).setSyncable(false);
 
     public Attributes() {}
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 09bcbc0ae36e4e69fee87a7e0c49acf496117a39..9e1c0d47dec605309e3ad8a680793665945138f3 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -7,16 +7,8 @@ import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.OptionalInt;
-import java.util.function.Predicate;
 import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
+import de.erethon.papyrus.DamageType;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -24,11 +16,7 @@ import net.minecraft.core.Direction;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.particles.ParticleTypes;
-import net.minecraft.nbt.CompoundTag;
-import net.minecraft.nbt.ListTag;
-import net.minecraft.nbt.NbtOps;
-import net.minecraft.nbt.NbtUtils;
-import net.minecraft.nbt.Tag;
+import net.minecraft.nbt.*;
 import net.minecraft.network.chat.ClickEvent;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
@@ -50,11 +38,7 @@ import net.minecraft.tags.EntityTypeTags;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.util.Unit;
-import net.minecraft.world.Container;
-import net.minecraft.world.Difficulty;
-import net.minecraft.world.InteractionHand;
-import net.minecraft.world.InteractionResult;
-import net.minecraft.world.MenuProvider;
+import net.minecraft.world.*;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffectUtil;
@@ -62,15 +46,6 @@ import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityAttachment;
 import net.minecraft.world.entity.EntityAttachments;
-import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.EquipmentSlot;
-import net.minecraft.world.entity.HumanoidArm;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MoverType;
-import net.minecraft.world.entity.Pose;
-import net.minecraft.world.entity.SlotAccess;
-import net.minecraft.world.entity.TamableAnimal;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.animal.Parrot;
@@ -88,12 +63,7 @@ import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ClickAction;
 import net.minecraft.world.inventory.InventoryMenu;
 import net.minecraft.world.inventory.PlayerEnderChestContainer;
-import net.minecraft.world.item.ElytraItem;
-import net.minecraft.world.item.ItemCooldowns;
-import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
-import net.minecraft.world.item.ProjectileWeaponItem;
-import net.minecraft.world.item.SwordItem;
+import net.minecraft.world.item.*;
 import net.minecraft.world.item.crafting.RecipeHolder;
 import net.minecraft.world.item.enchantment.EnchantmentEffectComponents;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
@@ -113,7 +83,6 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Scoreboard;
-import org.slf4j.Logger;
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.util.CraftVector;
@@ -126,6 +95,11 @@ import org.bukkit.event.entity.EntityKnockbackEvent;
 import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.player.PlayerDropItemEvent;
 import org.bukkit.event.player.PlayerVelocityEvent;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.function.Predicate;
 // CraftBukkit end
 
 public abstract class Player extends LivingEntity {
@@ -133,7 +107,7 @@ public abstract class Player extends LivingEntity {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final HumanoidArm DEFAULT_MAIN_HAND = HumanoidArm.RIGHT;
     public static final int DEFAULT_MODEL_CUSTOMIZATION = 0;
-    public static final int MAX_HEALTH = 20;
+    public static final int MAX_HEALTH = 500;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
     public static final int ENDER_SLOT_OFFSET = 200;
@@ -1265,6 +1239,15 @@ public abstract class Player extends LivingEntity {
         return this.level().noCollision(this, new AABB(axisalignedbb.minX + offsetX, axisalignedbb.minY - (double) f - 9.999999747378752E-6D, axisalignedbb.minZ + offsetZ, axisalignedbb.maxX + offsetX, axisalignedbb.minY, axisalignedbb.maxZ + offsetZ));
     }
 
+    // Papyrus - combat rework
+    public void attack(Entity target, DamageType type) {
+        if (!target.isAttackable()) {
+            return;
+        }
+        float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        target.hurt(damageSources().playerAttack(this), damage, type);
+    }
+
     public void attack(Entity target) {
         // Paper start - PlayerAttackEntityEvent
         boolean willAttack = target.isAttackable() && !target.skipAttackInteraction(this); // Vanilla logic
@@ -1503,7 +1486,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target);
+        this.attack(target, DamageType.PHYSICAL); // Papyrus
     }
 
     @io.papermc.paper.annotation.DoNotUse @Deprecated // Paper - Add PlayerShieldDisableEvent
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index d01b45a48d412e3cb591acee101730704574448a..72ae63e23229c62ee72b248dde1f82710b22d931 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -204,8 +204,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private final Set<UUID> unlistedEntities = new HashSet<>(); // Paper - Add Listing API for Player
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
-    private double health = 20;
-    private boolean scaledHealth = false;
+    private double health = 500; // Papyrus
+    private boolean scaledHealth = true; // Papyrus
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
